[SPARK-15487] [Web UI] Spark Master UI to reverse proxy Application and Workers UI

gurvindersingh opened the pull request:
## What changes were proposed in this pull request?

This pull request adds the functionality to enable accessing worker and application UI through master UI itself. Thus helps in accessing SparkUI when running spark cluster in closed networks e.g. Kubernetes. Cluster admin needs to expose only spark master UI and rest of the UIs can be in the private network, master UI will reverse proxy the connection request to corresponding resource. It adds the path for workers/application UIs as

WorkerUI: <http/https>://master-publicIP:<port>/target/workerID/
ApplicationUI: <http/https>://master-publicIP:<port>/target/appID/

This makes it easy for users to easily protect the Spark master cluster access by putting some reverse proxy e.g. https://github.com/bitly/oauth2_proxy
## How was this patch tested?

The functionality has been tested manually and there is a unit test too for testing access to worker UI with reverse proxy address.

@pwendell @bomeng @BryanCutler can you please review it, thanks.

Commit:
added support for reverse proxying request to workers and application UI. This helps in accessing SparkUI when running spark cluster in closed networks.

AmplabJenkins commented:
Can one of the admins verify this patch?

ajbozarth commented:
This looks good and I really like this added functionality, I'll check out your code and give it a test run as soon as I have time.

prb commented on core/src/main/scala/org/apache/spark/deploy/master/ui/ApplicationPage.scala:
```diff
@@ -77,7 +77,15 @@ private[ui] class ApplicationPage(parent: MasterWebUI) extends WebUIPage("app")
+                <li><strong>
+                  {
+                    if (parent.master.reverseProxy) {
+                      <a href={"/target/" + app.id + "/"}>Application Detail UI</a>
```
Should the `"/target/"` be a literal or a configuration value?

gurvindersingh replied:
By all means it can be a configuration value, but there were already some literal conventions eg. "app" so I thought of having it this way. But if community prefer the configured value, then I will add that.

prb commented on core/src/main/scala/org/apache/spark/ui/JettyUtils.scala:
```diff
@@ -186,6 +188,67 @@ private[spark] object JettyUtils extends Logging {
   }
 
+  /** Create a handler for proxying request to Workers and Application Drivers */
+  def createProxyHandler(
```
Always a little scary to see string manipulation code with no tests.

gurvindersingh replied:
I might have overlooked the tests for other handlers, can you point me to them if there is any. I can add test for this handler too then.

vanzin replied:
+1 to targeted tests of string manipulation code.

Move the string manipulation part into its own method, and write targeted unit tests for it.

gurvindersingh replied:
added a test under UISuite, let me know if that is not the right place for it.

ajbozarth commented on core/src/main/scala/org/apache/spark/deploy/master/Master.scala:
```diff
@@ -127,7 +128,14 @@ private[deploy] class Master(
     webUi.bind()
-    masterWebUiUrl = "http://" + masterPublicAddress + ":" + webUi.boundPort
+    if (reverseProxy) {
+      masterWebUiUrl = conf.get("spark.ui.reverseProxyUrl", null)
```
Is there a reason we need this config? Would there ever be a case where the proxy URL is different from the master URL? Also from my testing this config is completely ignored (as long as it exists)

gurvindersingh replied:
It is used in case when you are running spark master itself behind a proxy e.g. Oauth2 to provide authentication/authorization. Its to make sure "Back to Master"  link works when you are on workers UI.

ajbozarth commented on core/src/main/scala/org/apache/spark/deploy/master/Master.scala:
```diff
@@ -127,7 +128,14 @@ private[deploy] class Master(
+    if (reverseProxy) {
+      masterWebUiUrl = conf.get("spark.ui.reverseProxyUrl", null)
+      if (masterWebUiUrl == null) {
+       throw new SparkException("spark.ui.reverseProxyUrl must be provided")
```
When this is thrown the Spark Master still starts, but workings and applications can't connect to it.

gurvindersingh replied:
Updated the code now to remove the exception and use public address as default and if reverseproxyURL is given then override it. Should solve the issue you seeing.

Commit:
default to master public url, but if proxy url is given use that

gurvindersingh commented:
@ajbozarth what are the steps ahead in terms of moving this pr closer to being merged.

ajbozarth commented:
@srowen @zsxwing @tgravescs What do you guys think of @gurvindersingh proposal?

tnachen commented on docs/configuration.md:
```diff
@@ -598,6 +598,20 @@ Apart from these, the following properties are also available, and may be useful
+  <td><code>spark.ui.reverseProxyUrl</code></td>
+  <td>http://localhost:8080</td>
+  <td>
+    This is the URL where your proxy is running. Make sure this is a complete URL includeing scheme (http/https) and port to reach your proxy.
```
includeing -> including

tnachen commented on core/src/main/scala/org/apache/spark/ui/JettyUtils.scala:
```diff
@@ -186,6 +188,67 @@ private[spark] object JettyUtils extends Logging {
+        if (target.endsWith("/")) uri.setLength(uri.length() - 1)
+        val rest = path.substring(prefix.length())
+        if (!rest.isEmpty())
+        {
```
Move { to previous line

gurvindersingh replied:
Both are fixed now

Commit:
fixed syntax and a typo

tnachen commented on core/src/main/scala/org/apache/spark/SparkContext.scala:
```diff
@@ -502,6 +502,9 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
     _applicationAttemptId = taskScheduler.applicationAttemptId()
     _conf.set("spark.app.id", _applicationId)
+    if (_conf.getBoolean("spark.ui.reverseProxy", false)) {
+      System.setProperty("spark.ui.proxyBase", "/target/" + _applicationId)
```
I wonder if this is better named /proxy/app/applicationId for applications, and /proxy/worker/workerId, so it's more clear what the destination target is?

gurvindersingh replied:
I am not sure if it would have any benefit, as currently it will look like /target/worker-Id or /target/app-id so url does have information about whether it is application or worker. Regarding proxy or target name, I have no preference. As to me it does convey that we are accessing worker/app page. I can make it a configurable value if that's preferred.

tnachen replied:
ah I see, sorry I thought it will be just opaque ids, but worker- and app- prefixes looks fine to me.

gurvindersingh replied:
O, so what about target/proxy, target is fine or should I change it to proxy ?

gurvindersingh commented:
@tgravescs Yeah the proposal is only for standalone mode where worker & application UI is accessed through master UI now. Looking at the authn/z settings for standalone, I don't see this patch interfere with any of those. The addFilters() function set the ui.filters for '/*' path so it will apply for this case too. Beyond this let me know if I miss anything.

gurvindersingh commented:
@tgravescs @ajbozarth any update on this PR ?

ajbozarth commented:
I'll try and take a look at this tomorrow

ajbozarth commented:
I think changing `/target/` to `/proxy/` would make a clearer url.
@tgravescs what does the yarn proxy url look like? I don't use yarn outside of testing.

Overall LGTM though

Commit:
changed proxy url path from target to proxy to make it explicit that we are proxying

gurvindersingh commented:
Changed the path from `target` to `proxy` Let me know if there is anything else before this PR can be merged.

gurvindersingh commented:
@ajbozarth ping on this..

ajbozarth commented:
LGTM, @tgravescs this ready to merge?

gurvindersingh commented:
@ajbozarth it seems @tgravescs is busy, is anybody else then who you think could help us in merging this.

chandru-ops commented:
I'm trying to have nginx as reverse proxy for Spark dashboards which doesn't support app/driver logs now. Thank you @gurvindersingh . I will try this.

ajbozarth commented:
@srowen @JoshRosen could we get some traction on this? Thanks!

vanzin commented:
ok to test

SparkQA commented:
**[Test build #63232 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63232/consoleFull)** for PR 13950 at commit [`22f77c9`](https://github.com/apache/spark/commit/22f77c9829fc1f1aad938f83e30f4608b62b9e51).

SparkQA commented:
**[Test build #63232 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63232/consoleFull)** for PR 13950 at commit [`22f77c9`](https://github.com/apache/spark/commit/22f77c9829fc1f1aad938f83e30f4608b62b9e51).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/63232/
Test PASSed.

vanzin commented on core/pom.xml:
```diff
@@ -127,6 +127,16 @@
     </dependency>
     <dependency>
       <groupId>org.eclipse.jetty</groupId>
+      <artifactId>jetty-proxy</artifactId>
```
These new dependencies need to be added to the `copy-dependencies` invocation later in this file.

vanzin commented on core/src/main/scala/org/apache/spark/deploy/master/Master.scala:
```diff
@@ -753,6 +757,7 @@ private[deploy] class Master(
     workers += worker
     idToWorker(worker.id) = worker
     addressToWorker(workerAddress) = worker
+    if (reverseProxy) webUi.addProxyTargets(worker.id, worker.webUiAddress)
```
style: please use the "long form" of if statements (unless they're used as a ternary operator)

vanzin commented on core/src/main/scala/org/apache/spark/ui/JettyUtils.scala:
```diff
@@ -186,6 +188,66 @@ private[spark] object JettyUtils extends Logging {
+
+        rewrittenURI.toString();
+      }
+      override def filterServerResponseHeader(
```
nit: add empty line between methods

vanzin commented on core/src/main/scala/org/apache/spark/ui/JettyUtils.scala:
```diff
@@ -186,6 +188,66 @@ private[spark] object JettyUtils extends Logging {
+      target: String): ServletContextHandler = {
+    val servlet = new ProxyServlet {
+      override def rewriteTarget(request: HttpServletRequest): String = {
+        val path = request.getRequestURI();
```
nit: no semi-colons in scala code

vanzin commented on core/src/main/scala/org/apache/spark/ui/JettyUtils.scala:
```diff
@@ -186,6 +188,66 @@ private[spark] object JettyUtils extends Logging {
+        rewrittenURI.toString();
+      }
+      override def filterServerResponseHeader(
+        clientRequest: HttpServletRequest,
```
style: multi-line argument lists need to be indented one extra level.

vanzin commented on core/src/main/scala/org/apache/spark/ui/JettyUtils.scala:
```diff
@@ -186,6 +188,66 @@ private[spark] object JettyUtils extends Logging {
+        if (query != null) {
+          // Is there at least one path segment ?
+          val separator = "://"
+          if (uri.indexOf("/", uri.indexOf(separator) + separator.length()) < 0) {
```
Any reason not to use `java.net.URI` here instead of manual parsing code?

vanzin commented on core/src/main/scala/org/apache/spark/ui/JettyUtils.scala:
```diff
@@ -186,6 +188,66 @@ private[spark] object JettyUtils extends Logging {
+        if (!path.startsWith(prefix)) return null
+
+        val uri = new StringBuilder(target)
+        if (target.endsWith("/")) uri.setLength(uri.length() - 1)
```
Seems like it would be better to normalize `target` and `prefix` before instantiating the servlet, so that they conform to whatever you expect, and avoid these extra checks on every invocation of the method.

e.g., make sure `target` does not end with a slash, and `prefix` does.

vanzin commented on core/src/main/scala/org/apache/spark/ui/JettyUtils.scala:
```diff
@@ -186,6 +188,66 @@ private[spark] object JettyUtils extends Logging {
+        serverResponse: Response,
+        headerName: String,
+        headerValue: String): String = {
+        if (headerName.equalsIgnoreCase("location")) {
```
It would also be nice to have targeted unit tests for this code.

vanzin commented:
One idea is that if you use `spark.ui.proxyBase`, as YARN does, you don't need the proxy servlet to modify the HTML returned to the clients. But then you'd need something like Hadoop's `AmIpFilter` installed so that the app UI redirects to the master...

Commit:
fixed syntax and added test for proxyURI

gurvindersingh commented:
@vanzin addressed most of your comments. Wondering if the test for location header is really needed, as it is a very simple two if checks. Let me know if you feel strongly about that or any other changes are required.

SparkQA commented:
**[Test build #63731 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63731/consoleFull)** for PR 13950 at commit [`222fd1d`](https://github.com/apache/spark/commit/222fd1dc5944a851f4aa267a01c2a015b12b3ec9).

SparkQA commented:
**[Test build #63731 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63731/consoleFull)** for PR 13950 at commit [`222fd1d`](https://github.com/apache/spark/commit/222fd1dc5944a851f4aa267a01c2a015b12b3ec9).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/63731/
Test PASSed.

vanzin commented on core/src/main/scala/org/apache/spark/deploy/master/ui/ApplicationPage.scala:
```diff
@@ -100,19 +108,21 @@ private[ui] class ApplicationPage(parent: MasterWebUI) extends WebUIPage("app")
   }
 
   private def executorRow(executor: ExecutorDesc): Seq[Node] = {
+    var workerUrlRef = executor.worker.webUiAddress
```
```
 val workerUrlRef = if (parent.master.reverseProxy) ... else ...
```

vanzin commented on core/src/main/scala/org/apache/spark/deploy/master/ui/MasterPage.scala:
```diff
@@ -244,7 +254,12 @@ private[ui] class MasterPage(parent: MasterWebUI) extends WebUIPage("") {
-      <td>{driver.worker.map(w => <a href={w.webUiAddress}>{w.id.toString}</a>).getOrElse("None")}
+      <td>{driver.worker.map(w =>
+        if (parent.master.reverseProxy) {
+          <a href={"/proxy/" + w.id.toString + "/"}>{w.id.toString}</a>
```
nit: `toString` is unnecessary.

Also, interpolation would look better: `s"/proxy/${w.id}/"`

vanzin commented on core/src/main/scala/org/apache/spark/deploy/master/ui/MasterWebUI.scala:
```diff
@@ -48,6 +53,21 @@ class MasterWebUI(
   }
+
+  def addProxyTargets(id: String, target: String): Unit = {
+    var endTarget: String = target
```
`val endTarget = target.stripSuffix("/")`

vanzin commented on core/src/main/scala/org/apache/spark/deploy/master/ui/MasterWebUI.scala:
```diff
@@ -48,6 +53,21 @@ class MasterWebUI(
+  }
+
+  def removeProxyTargets(id: String): Unit = {
+    detachHandler(proxyHandlers.get(id).get)
```
`proxyHandlers.remove(id).foreach(detachHandler)`

vanzin commented on core/src/main/scala/org/apache/spark/ui/JettyUtils.scala:
```diff
@@ -186,6 +188,50 @@ private[spark] object JettyUtils extends Logging {
+    val servlet = new ProxyServlet {
+      override def rewriteTarget(request: HttpServletRequest): String = {
+        val rewrittenURI = createProxyURI(
+                              prefix,
```
nit: style. If arguments don't fit in one line, they should be indented two levels (= 4 spaces).

vanzin commented on core/src/main/scala/org/apache/spark/ui/JettyUtils.scala:
```diff
@@ -186,6 +188,50 @@ private[spark] object JettyUtils extends Logging {
+        if (!validateDestination(rewrittenURI.getHost(), rewrittenURI.getPort())) {
+          return null
+        }
+        rewrittenURI.toString();
```
nit: no semi-colons.

vanzin commented on core/src/test/scala/org/apache/spark/deploy/master/MasterSuite.scala:
```diff
@@ -157,6 +157,34 @@ class MasterSuite extends SparkFunSuite
+          val JString(workerId) = workerSummaryJson \ "id"
+          val url = s"http://localhost:${localCluster.masterWebUIPort}/proxy/${workerId}/json"
+          val workerResponse = parse(Source.fromURL(url)
+            .getLines().mkString("\n"))
```
nit: seems like this fits in the previous line

vanzin commented:
> Wondering if the test for location header is really needed

Always a good idea to have tests, even if you feel the code is simple. Then if someone needs to change the code later, the tests might catch anything wrong.

Also, it seems you implemented the proxyBase approach I suggested. Can you confirm that, when proxying is enabled, you can still look at the app's UI by going to the app's web server directly, and not just through the master UI?

(In YARN you get redirected to the "master" - the YARN RM in that case - when going to the app's web server, but here you don't have any code to do that, so the UI should still work when setting proxyBase.)

Commit:
added test for location header and fixed more syntax

SparkQA commented:
**[Test build #63847 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63847/consoleFull)** for PR 13950 at commit [`d56238d`](https://github.com/apache/spark/commit/d56238d977066f6ef4c7363c658f91700c255acf).

gurvindersingh commented:
added test for location header too and fixed syntax too ( I am bit new to scala, so thanks for bearing it with me). 

Regarding application UI being accessible when accessed directly, in the current code use case it will not work. As having master to reverse proxy the application and worker ui indicate that you don't have access to application UI directly. So I am not sure as this is a valid use case in this scenario. 

If you think you do need this to be supported, can you give me use case and point me as how its is done in yarn mode, I can add code for standalone proxy mode too then.

SparkQA commented:
**[Test build #63847 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63847/consoleFull)** for PR 13950 at commit [`d56238d`](https://github.com/apache/spark/commit/d56238d977066f6ef4c7363c658f91700c255acf).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/63847/
Test FAILed.

vanzin commented:
> I can add info about app UI not being accessible directly when reverse proxy is enabled in the documentation, so users do know about this.

I think that's the bare minimum, since users might be surprised by the behavior. As for the behavior itself, it seems inline with the goal of the patch (where users wouldn't have direct access to the app's UI, only through the master). But one thing is that it requires both the master and all the applications to share that config - and that needs to be clearly documented. In cluster mode, it would even be nice if the master / worker daemons set that config when launching the app.

The YARN code is in `ApplicationMaster` (look for `def addAmIpFilter`). But that relies on a YARN class that might not be available to the master / worker daemons, so you can't really use it here.

Commit:
added docs with warning about reverse proxy being global to cluster and restricting direct access to application UI

SparkQA commented:
**[Test build #63859 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63859/consoleFull)** for PR 13950 at commit [`6499d2e`](https://github.com/apache/spark/commit/6499d2e6d7cff4f6186fa94ef1c9ea575c4fe5a9).

SparkQA commented:
**[Test build #63860 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63860/consoleFull)** for PR 13950 at commit [`8254a2d`](https://github.com/apache/spark/commit/8254a2da9d7ef3da223c561edc212a6027f0ea2a).

gurvindersingh commented:
OK, I have added docs about using this with caution, as it will disable direct application UI access. Regarding redirecting automatically to master, I faced issue as during application UI I couldn't access masterwebuiUrl and applicationId, otherwise I could simply added the redirect as

```
if (reverseProxy) {
  attachHandler(createRedirectHandler("/", s"${masterwebuiUrl}/proxy/${appId}/jobs", basePath = basePath))
} else {
  attachHandler(createRedirectHandler("/", "/jobs/", basePath = basePath))
}
```

In any case, user now clearly sees in the documentation as when reverse proxy is enabled use master public URL to access all the UIs.

SparkQA commented:
**[Test build #63859 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63859/consoleFull)** for PR 13950 at commit [`6499d2e`](https://github.com/apache/spark/commit/6499d2e6d7cff4f6186fa94ef1c9ea575c4fe5a9).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/63859/
Test PASSed.

SparkQA commented:
**[Test build #63860 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63860/consoleFull)** for PR 13950 at commit [`8254a2d`](https://github.com/apache/spark/commit/8254a2da9d7ef3da223c561edc212a6027f0ea2a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/63860/
Test PASSed.

ajbozarth commented:
Leaving the code review to vanzin I checked out and ran your updated code and found a few issues. 

You should probably merge the latest code, the ui has been undergoing a handful of updates recently that you should check for compatibility with.

On the executors page the links for logs don't use the proxy.

When spark-shell is run it still says "Spark context Web UI available at http://localhost:4040" even though you have disabled that with the proxy.

SparkQA commented:
**[Test build #63907 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63907/consoleFull)** for PR 13950 at commit [`035478d`](https://github.com/apache/spark/commit/035478d80dd98fb902737f2c4c2039f475959848).

SparkQA commented:
**[Test build #63909 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63909/consoleFull)** for PR 13950 at commit [`624af2e`](https://github.com/apache/spark/commit/624af2eaea7789184fc4e7a1772d55aeb956a70f).

SparkQA commented:
**[Test build #63907 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63907/consoleFull)** for PR 13950 at commit [`035478d`](https://github.com/apache/spark/commit/035478d80dd98fb902737f2c4c2039f475959848).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/63907/
Test PASSed.

SparkQA commented:
**[Test build #63909 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63909/consoleFull)** for PR 13950 at commit [`624af2e`](https://github.com/apache/spark/commit/624af2eaea7789184fc4e7a1772d55aeb956a70f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/63909/
Test PASSed.

Commit:
print information about accessing UIs when running in proxy mode

vanzin commented on core/src/main/scala/org/apache/spark/deploy/master/ui/ApplicationPage.scala:
```diff
@@ -100,19 +108,24 @@ private[ui] class ApplicationPage(parent: MasterWebUI) extends WebUIPage("app")
   }
 
   private def executorRow(executor: ExecutorDesc): Seq[Node] = {
+    var workerUrlRef = if (parent.master.reverseProxy) {
```
Indentation here is odd.

```
val foo = if (...) {
  // code
} else {
  // code
}
```

SparkQA commented:
**[Test build #63932 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63932/consoleFull)** for PR 13950 at commit [`71bba1f`](https://github.com/apache/spark/commit/71bba1f7a4cb470b9c7482bb1e1313cd7fdd359d).

gurvindersingh commented:
I have rebased the code on current master and it seems in the new master their are few more places where executor logs are accessible. But they all seems to work for me. I tested with running spark-shell and access the logs from executor tab, Stages tab. Also thread dump works too. Let me know if there are some other places which I missed.

Regarding printing information for accessing UI, I have added message on master, worker and changed print message when running spark-shell. I couldn't give exact URL in case of spark-shell as I   didn't find way to access masterwebuiUrl from there. If it is accessible somehow let me know, I will add.

vanzin commented on core/src/main/scala/org/apache/spark/deploy/master/ui/MasterPage.scala:
```diff
@@ -176,7 +176,13 @@ private[ui] class MasterPage(parent: MasterWebUI) extends WebUIPage("") {
       <td>
-        <a href={worker.webUiAddress}>{worker.id}</a>
+        {
+          if (parent.master.reverseProxy) {
```
This pattern is being used in a lot of places and it's kinda becoming noisy... perhaps a better approach would be to have a method somewhere that takes both the proxied and non-proxied versions, and returns which one to use. e.g.

    def href(normal: String, proxied: String): String = if (reverseProxy) s"/proxy/$proxied" else normal

vanzin commented on core/src/main/scala/org/apache/spark/ui/JettyUtils.scala:
```diff
@@ -186,6 +188,50 @@ private[spark] object JettyUtils extends Logging {
+    val servlet = new ProxyServlet {
+      override def rewriteTarget(request: HttpServletRequest): String = {
+        val rewrittenURI = createProxyURI(
+                                prefix,
```
indentation is wrong, should be 4 spaces (= 2 indent levels).

vanzin commented on core/src/main/scala/org/apache/spark/ui/JettyUtils.scala:
```diff
@@ -186,6 +188,50 @@ private[spark] object JettyUtils extends Logging {
+          headerValue: String): String = {
+        if (headerName.equalsIgnoreCase("location")) {
+          val newHeader = createProxyLocationHeader(
+                              prefix,
```
indentation

vanzin commented on core/src/main/scala/org/apache/spark/ui/JettyUtils.scala:
```diff
@@ -186,6 +188,50 @@ private[spark] object JettyUtils extends Logging {
+                              headerValue,
+                              clientRequest,
+                              serverResponse.getRequest().getURI())
+          if (newHeader != null) return newHeader
```
```
if (...) {
   return ...
}
```

SparkQA commented:
**[Test build #63933 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63933/consoleFull)** for PR 13950 at commit [`5139842`](https://github.com/apache/spark/commit/5139842e16a5d39b240cbc1e71fa8e4ece46242b).

vanzin commented on core/src/test/scala/org/apache/spark/ui/UISuite.scala:
```diff
@@ -190,6 +193,42 @@ class UISuite extends SparkFunSuite {
+    when(clientRequest.getScheme()).thenReturn("http")
+    when(clientRequest.getHeader("host")).thenReturn("localhost:8080")
+    var newHeader = JettyUtils.createProxyLocationHeader(
+                                  prefix,
```
indentation

vanzin commented on core/src/test/scala/org/apache/spark/ui/UISuite.scala:
```diff
@@ -190,6 +193,42 @@ class UISuite extends SparkFunSuite {
+    assert(newHeader.toString().equals("http://localhost:8080/proxy/worker-id/jobs"))
+    headerValue = "http://localhost:4041/jobs"
+    newHeader = JettyUtils.createProxyLocationHeader(
+                                  prefix,
```
indentation

vanzin commented on docs/configuration.md:
```diff
@@ -627,6 +627,20 @@ Apart from these, the following properties are also available, and may be useful
+  <td><code>spark.ui.reverseProxy</code></td>
+  <td>false</td>
+  <td>
+    To enable running Spark Master, worker and application UI behined a reverse proxy. In this mode, Spark master will reverse proxy the worker and application UIs to enable access. Use it with caution, as worker and application UI will not be accessible directly, you will only be able to access it through spark master/proxy public URL. This setting affects all the workers and application UIs run ...
```
"Worker"
"behind"
"Master"
"Spark"

vanzin commented on docs/configuration.md:
```diff
@@ -627,6 +627,20 @@ Apart from these, the following properties are also available, and may be useful
+</tr>
+<tr>
+  <td><code>spark.ui.reverseProxyUrl</code></td>
+  <td>http://localhost:8080</td>
```
There is no default. The default is the master's address computed at runtime.

vanzin commented on repl/scala-2.11/src/main/scala/org/apache/spark/repl/SparkILoop.scala:
```diff
@@ -43,7 +43,11 @@ class SparkILoop(in0: Option[BufferedReader], out: JPrintWriter)
+          if (_sc.getConf.getBoolean("spark.ui.reverseProxy", false)) {
+            println(s"Spark Context Web UI is available at Spark Master Public URL")
+          } else {
+            _sc.uiWebUrl.foreach(webUrl => println(s"Spark context Web UI available at ${webUrl}"))
```
nit: `.foreach { webUrl => ... }`

vanzin commented:
The code looks alright pending some minor cleanup, but it would be nice if someone more active in standalone could take a look and potentially help test this. Maybe @JoshRosen or @zsxwing ?

SparkQA commented:
**[Test build #63932 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63932/consoleFull)** for PR 13950 at commit [`71bba1f`](https://github.com/apache/spark/commit/71bba1f7a4cb470b9c7482bb1e1313cd7fdd359d).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/63932/
Test FAILed.

SparkQA commented:
**[Test build #63936 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63936/consoleFull)** for PR 13950 at commit [`c5b4ae9`](https://github.com/apache/spark/commit/c5b4ae970689b3fdfa262d169f0cdb85abbdf03b).

gurvindersingh commented:
addressed all of your comments now and the test which is failing is not related to my change.

SparkQA commented:
**[Test build #63933 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63933/consoleFull)** for PR 13950 at commit [`5139842`](https://github.com/apache/spark/commit/5139842e16a5d39b240cbc1e71fa8e4ece46242b).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/63933/
Test PASSed.

SparkQA commented:
**[Test build #63936 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63936/consoleFull)** for PR 13950 at commit [`c5b4ae9`](https://github.com/apache/spark/commit/c5b4ae970689b3fdfa262d169f0cdb85abbdf03b).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/63936/
Test PASSed.

ajbozarth commented:
@gurvindersingh I checked out your latest code and ran it, stdout and stderr log links on StagePage and ExecutorsPage are still not using the proxy url for me.

Addressing the shell start up, you have the proxy url in the conf (since you added it) so you have access to it there just like your proxy check

gurvindersingh commented:
@ajbozarth That is strange. Here is the steps I used to test and its working on my side

```
1. git clone https://github.com/apache/spark.git
2. edit .git/config to allow fetching pr and fetch pr
3. git checkout origin/pr/13950
4. build/sbt clean package
5. duplicate spark-defaults.conf.template to conf and add this option spark.ui.reverseProxy   true
6. Run Spark local cluster as
   - ./bin/spark-class org.apache.spark.deploy.master.Master
   - ./bin/spark-class org.apache.spark.deploy.worker.Worker spark://<spark-master-ip>:7077
   - ./bin/spark-shell --master=spark://<spark-master-ip>:7077
7. Ran sample task sc.parallelize(1 to 1000).reduce(_+_)
8 Go to http://<spark-master-ip>:8080
9. There you can go to application UI by going to application page and it is proxied from master ip as http://<spark-master-ip>:8080/proxy/app-id/
10. Similarly on Stagetab and executors tab, I am able to access the links url as http://<spark-master-ip>:8080/proxy/worker-id/logPage/....
```
Now To test running the spark master itself behind another proxy, stop the current cluster.
```
1. Start a local proxy any tool you prefer. 
  - e.g. ssh -L 8000:localhost:8080 localhost
2. Enter new option in conf file as spark.ui.reverseProxyUrl    http://localhost:8000 (enter the proxy address here)
3. Start the cluster similar to step above and run sample task
4. Go to http://localhost:8000/ and you can access the master, worker and application UI all through this URL without need to go to their UIs directly.
```

Regarding access to Master url for printing info for spark-shell, the reverseProxyUrl is not always set. As it is not necessary to run master itself behind reverse proxy. So can not use that to print information reliably about app UI url.

ajbozarth commented:
@gurvindersingh ah there's the issue, when your Worker is on the same system as your Master the stdout/stderr links use the proxy but if the Worker is on another system the links still go to that system's ip address directly.

As for `reverseProxyUrl` in the shell, I was thinking that you could use it *if* it's set and leave the text as is if it's not.

SparkQA commented:
**[Test build #64031 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64031/consoleFull)** for PR 13950 at commit [`032ac0e`](https://github.com/apache/spark/commit/032ac0ecfb14fba2a0d2872b406993243d39ca8b).

gurvindersingh commented:
@ajbozarth I have changed the information part when `reverseproxyUrl` is defined. Regarding accessing worker logs it is still working for me when worker, master and application are on different IP. Steps I used to test

```
1. Run 3 docker container as
 - docker run -it -v /path/to/compiled/spark/with/patch:/spark java bash (I used the same folder in above steps)
2. In one docker container start spark master, another worker and third spark-shell. Remember to give the spark-master IP as it is different IP for all of them now.
3. Ran sample task in shell
4. Go to  http://<spark-master-ip>:8080
9. There you can go to application UI by going to application page and it is proxied from master ip as http://<spark-master-ip>:8080/proxy/app-id/
10. Similarly on Stagetab and executors tab, I am able to access the links url as http://<spark-master-ip>:8080/proxy/worker-id/logPage/....
```

I am not able to see the issue, can you details the step to reproduce.

SparkQA commented:
**[Test build #64031 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64031/consoleFull)** for PR 13950 at commit [`032ac0e`](https://github.com/apache/spark/commit/032ac0ecfb14fba2a0d2872b406993243d39ca8b).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/64031/
Test PASSed.

ajbozarth commented:
Ah I know what the issue is now. Your implementation needs the `spark.ui.reverseProxy` conf set on every worker node to function. Since conf files aren't included in tar balls my workers didn't have those configs set (I used tar ball distributions of spark on my workers and a git clone on my master). From my experience many spark clusters don't set conf files on their workers, you may want to instead pass the value of `spark.ui.reverseProxy` to each worker in some way rather than looking it up in the workers confs (this happens in `ExecutorRunner.scala`).
@vanzin what's your opinion on this? Is it ok to assume the confs are set for each slave or should you assume they aren't?

vanzin commented:
I'm not too familiar with standalone deployments, but my understanding is that workers do not fetch configuration from the master, they just use their local copy. So the admin would have to set that option on every worker for the proxy to work.

BTW that's the same for applications here - the code assumes that the config will be set when the application starts, otherwise proxying will not be enabled for that application, and the links on the master UI will not work.

ajbozarth commented:
Yeah I just ran a few tests and the local spark, the master and the works all need the conf otherwise the web ui will behave oddly. I tested with the Master configured and the application not and the master tries to use the proxy but it doesn't work, but the app ui is still accessible at it's normal location. Inversely if the app is configured but the master isn't then the ui doesn't exist anywhere because it's not on the web ui port and the master isn't set up for proxy. As for the works not configured, that shows as log links not using the proxy (also note the worker UI's are available at their normal location whether or not proxy is on).

ajbozarth commented:
@srowen @JoshRosen We're having some discussion about best practices with configs for stand alone web ui, any input?

gurvindersingh commented:
As I mentioned in my last comment, the current code assume that the conf value is set on all components. E.g. Otherwise how you expect reverseProxyUrl to work in spark-shell, every `SparkContext`has its own conf.  As worker and application has reference to master but seems only using RpcEndpoint, so not possible to access the variable (as they usually are on different machines). So if you can suggest any way of passing the proxy and proxyUrl config to worker and application, I can implement it. I have no strong opinion on this, for my use case I use docker images and this config value is set in the base image from where master, worker and driver are built so they all have this conf value in them. 

Anyway as I said, point me in the direction of sharing the config/other method to access proxy & proxyUrl on wokers and applications, I will add support for the same.

gurvindersingh commented:
@ajbozarth @vanzin I have added docs to make sure the user set this setting on all the components. I think the possibility of having the option to specify only at master can be added in subsequent pull request. The current PR clearly states when and how to use this feature. Let me know your views on it.

SparkQA commented:
**[Test build #64348 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64348/consoleFull)** for PR 13950 at commit [`86f3f84`](https://github.com/apache/spark/commit/86f3f8417aca840f70d9444abd6654c29494c0e3).

SparkQA commented:
**[Test build #64348 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64348/consoleFull)** for PR 13950 at commit [`86f3f84`](https://github.com/apache/spark/commit/86f3f8417aca840f70d9444abd6654c29494c0e3).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/64348/
Test FAILed.

ajbozarth commented:
As long as everyone is ok with it and it's well documented this looks good to me, can a committer with experience with the Standalone Web UI take a look? @JoshRosen @srowen @zsxwing

SparkQA commented:
**[Test build #64490 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64490/consoleFull)** for PR 13950 at commit [`a524a8b`](https://github.com/apache/spark/commit/a524a8b97cf77d44ef5e7826300a93c306507f9d).

SparkQA commented:
**[Test build #64490 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64490/consoleFull)** for PR 13950 at commit [`a524a8b`](https://github.com/apache/spark/commit/a524a8b97cf77d44ef5e7826300a93c306507f9d).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/64490/
Test PASSed.

gurvindersingh commented:
@ajbozarth @vanzin @JoshRosen it seems to be stalled, any update ?

vanzin commented on core/src/main/scala/org/apache/spark/deploy/master/ui/ApplicationPage.scala:
```diff
@@ -100,19 +105,21 @@ private[ui] class ApplicationPage(parent: MasterWebUI) extends WebUIPage("app")
   }
 
   private def executorRow(executor: ExecutorDesc): Seq[Node] = {
+    var workerUrlRef = UIUtils.makeHref(parent.master.reverseProxy,
```
val

vanzin commented on core/src/main/scala/org/apache/spark/deploy/master/ui/ApplicationPage.scala:
```diff
@@ -77,7 +77,12 @@ private[ui] class ApplicationPage(parent: MasterWebUI) extends WebUIPage("app")
               if (!app.isFinished) {
-                <li><strong><a href={app.desc.appUiUrl}>Application Detail UI</a></strong></li>
+                <li><strong>
+                  {
```
Are the braces really necessary? (Here and in other parts of your change too.)

vanzin commented on core/src/main/scala/org/apache/spark/ui/JettyUtils.scala:
```diff
@@ -332,6 +375,47 @@ private[spark] object JettyUtils extends Logging {
+      targetUri: URI): String = {
+    val toReplace = targetUri.getScheme() + "://" + targetUri.getAuthority()
+    if (headerValue.startsWith(toReplace)) {
+      return clientRequest.getScheme() + "://" + clientRequest.getHeader("host") +
```
In these cases, instead of `return`, it's preferred to just do:

```
if (foo) {
  bar
} else {
  baz
}
```

vanzin commented on core/src/main/scala/org/apache/spark/ui/UIUtils.scala:
```diff
@@ -510,4 +510,15 @@ private[spark] object UIUtils extends Logging {
+  * reverse proxy mode or not.
+  */
+  def makeHref(proxy: Boolean, id: String, origHref: String): String = {
+    if (proxy) {
```
Similar to previous comment.

vanzin commented on core/src/test/scala/org/apache/spark/ui/UISuite.scala:
```diff
@@ -190,6 +193,36 @@ class UISuite extends SparkFunSuite {
+    val target = "http://localhost:8081"
+    val path = "/proxy/worker-id/json"
+    var rewrittenURI = JettyUtils.createProxyURI(prefix, target, path, null)
+    assert(rewrittenURI.toString().equals("http://localhost:8081/json"))
```
use `===` instead of `equals`

vanzin commented on core/src/test/scala/org/apache/spark/ui/UISuite.scala:
```diff
@@ -190,6 +193,36 @@ class UISuite extends SparkFunSuite {
+    rewrittenURI = JettyUtils.createProxyURI(prefix, target, "/proxy/worker-id", null)
+    assert(rewrittenURI.toString().equals("http://localhost:8081"))
+    rewrittenURI = JettyUtils.createProxyURI(prefix, target, "/proxy/worker-noid/json", null)
+    assert(rewrittenURI == null)
```
`===`

vanzin commented on docs/configuration.md:
```diff
@@ -635,6 +635,20 @@ Apart from these, the following properties are also available, and may be useful
+  <td><code>spark.ui.reverseProxy</code></td>
+  <td>false</td>
+  <td>
+    Enable running Spark Master as reverse proxy for worker and application UI. In this mode, Spark master will reverse proxy the worker and application UIs to enable access without requiring direct access to their UIs. Use it with caution, as worker and application UI will not be accessible directly, you will only be able to access it through spark master/proxy public URL. This setting affects a ...
```
everywhere: "UIs" (since you're saying "worker an application")

"...without requiring direct access to their hosts."

"...you will only be able to access them..."

"...workers, drivers and masters." (Yes there can be more than one.)

vanzin commented:
A few minor things. Looks ok, but I'd still like from someone who's more active in maintaining standalone mode to take a look.

If there's no comments in a few days I'll push it after the comments are addressed.

zsxwing commented on core/src/test/scala/org/apache/spark/ui/UISuite.scala:
```diff
@@ -190,6 +193,36 @@ class UISuite extends SparkFunSuite {
     }
   }
 
+  test("verify proxy rewrittenURI") {
```
Could you add a test for encoded string in the url, such as `%2F`, `%F0%9F%98%84` (the 😄 emoji) ?

gurvindersingh commented:
@vanzin addressed all your comments and added two extra checks too in test as requested. Let me  know if there are any more comments.

SparkQA commented:
**[Test build #65099 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65099/consoleFull)** for PR 13950 at commit [`407c1a0`](https://github.com/apache/spark/commit/407c1a0b5e0d229e1f0a92bf9afeb58e03506207).

SparkQA commented:
**[Test build #65099 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65099/consoleFull)** for PR 13950 at commit [`407c1a0`](https://github.com/apache/spark/commit/407c1a0b5e0d229e1f0a92bf9afeb58e03506207).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/65099/
Test PASSed.

zsxwing commented on core/src/main/scala/org/apache/spark/deploy/master/ui/MasterWebUI.scala:
```diff
@@ -34,6 +38,7 @@ class MasterWebUI(
 
   val masterEndpointRef = master.self
   val killEnabled = master.conf.getBoolean("spark.ui.killEnabled", true)
+  val proxyHandlers = new HashMap[String, ServletContextHandler]
```
nit: private[this]

zsxwing commented:
LGTM expect one nit!

SparkQA commented:
**[Test build #65112 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65112/consoleFull)** for PR 13950 at commit [`9f6862e`](https://github.com/apache/spark/commit/9f6862ed579cf4edc965e1a9835a79d6aa4c6d0b).

SparkQA commented:
**[Test build #65112 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65112/consoleFull)** for PR 13950 at commit [`9f6862e`](https://github.com/apache/spark/commit/9f6862ed579cf4edc965e1a9835a79d6aa4c6d0b).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/65112/
Test PASSed.

zsxwing commented:
LGTM. Merging to master! Thanks!
